File: ios/Sources/SIT/Views/Tickle/TickleActionSheet.swift
import SwiftUI
import SwiftData

/// Connect-action sheet shown when a tickle row is tapped (TIC-36).
/// Surfaces the contact's available channels (Compose message / Call / Email) plus
/// the secondary tickle actions (Mark done / Snooze / Edit). Compose opens the in-app
/// Compose screen (templates); Call/Email hand off to system composers. No contact
/// data leaves the device.
///
/// All actions dismiss the sheet; Compose and Edit additionally signal the parent
/// (via `onCompose` / `onEdit`) to present a follow-up sheet once this one is gone,
/// which avoids the "present while dismissing" race.
///
/// On iPad (TIC-46) the same content renders in a `NavigationSplitView` detail pane
/// rather than a modal sheet. Pass `dismissesOnAction: false` there so taps don't try
/// to dismiss a pane that was never presented — the parent drives the pane via the
/// callbacks and its own selection state instead.
struct TickleActionSheet: View {
    let reminder: TickleReminder
    var warmth: Warmth = .subtle
    /// When false (iPad detail pane), actions run their callback/URL but never call
    /// `dismiss()` — there's no sheet to dismiss. Defaults to true (phone bottom sheet).
    var dismissesOnAction: Bool = true
    let onCompose: () -> Void
    let onMarkDone: () -> Void
    /// Snooze the reminder by the chosen number of days (TIC-87 duration choice).
    let onSnooze: (Int) -> Void
    let onEdit: () -> Void

    @Environment(\.dismiss) private var dismiss
    @Environment(\.openURL) private var openURL

    /// Drives the snooze-duration confirmation dialog (1 day / 3 days / 1 week).
    @State private var showingSnoozeOptions = false

    /// Dismiss only when presented as a sheet; a no-op in the iPad detail pane.
    private func dismissIfNeeded() { if dismissesOnAction { dismiss() } }

    private var palette: WarmPalette { WarmTheme.palette(for: warmth) }

    private var category: WarmCategory {
        if let contact = reminder.contact { return WarmCategory.resolve(for: contact) }
        if let group = reminder.group, let cat = WarmCategory.from(groupId: group.id) { return cat }
        return .community
    }

    private var phones: [String] { reminder.contact?.phoneNumbers ?? [] }
    private var emails: [String] { reminder.contact?.emails ?? [] }
    /// Compose requires both a number to text and a device that can send text
    /// (false on iPad/Simulator/devices without Messages configured).
    private var canCompose: Bool { !phones.isEmpty && MessageComposerView.canSendMessages() }

    private var displayName: String {
        if let contact = reminder.contact { return contact.fullName }
        if let group = reminder.group { return group.displayName }
        return String(localized: "tickleRow.unknown")
    }

    /// Estimated content height, kept in sync with the rows rendered below so the
    /// parent's `.presentationDetents([.height(...)])` snugly fits the sheet.
    static func estimatedHeight(for reminder: TickleReminder) -> CGFloat {
        let phones = reminder.contact?.phoneNumbers ?? []
        let emails = reminder.contact?.emails ?? []
        var rows = 3 // mark done, snooze, edit
        if !phones.isEmpty && MessageComposerView.canSendMessages() { rows += 1 } // compose
        if !phones.isEmpty { rows += 1 } // call
        if !emails.isEmpty { rows += 1 } // email
        let header: CGFloat = 92
        let rowHeight: CGFloat = 56
        let bottomPadding: CGFloat = 28
        return header + CGFloat(rows) * rowHeight + bottomPadding
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            header

            if canCompose {
                actionRow(icon: "square.and.pencil",
                          title: String(localized: "tickleAction.compose", defaultValue: "Compose message"),
                          tint: category.palette.accent) {
                    onCompose()
                    dismissIfNeeded()
                }
            }

            if !phones.isEmpty {
                actionRow(icon: "phone.fill",
                          title: String(localized: "warm.contact.call", defaultValue: "Call"),
                          tint: category.palette.accent) {
                    if let phone = phones.first,
                       let url = URL(string: "tel:\(phone.filter { $0.isNumber || $0 == "+" })") {
                        openURL(url)
                    }
                    dismissIfNeeded()
                }
            }

            if !emails.isEmpty {
                actionRow(icon: "envelope.fill",
                          title: String(localized: "contact.section.email", defaultValue: "Email"),
                          tint: category.palette.accent) {
                    if let email = emails.first, let url = URL(string: "mailto:\(email)") {
                        openURL(url)
                    }
                    dismissIfNeeded()
                }
            }

            WarmRowDivider(warmth: warmth)
                .padding(.vertical, 6)

            actionRow(icon: "checkmark.circle.fill",
                      title: String(localized: "common.done", defaultValue: "Done"),
                      tint: palette.ink2) {
                onMarkDone()
                dismissIfNeeded()
            }

            actionRow(icon: "zzz",
                      title: String(localized: "tickleList.action.snooze", defaultValue: "Snooze"),
                      tint: palette.ink2) {
                showingSnoozeOptions = true
            }

            actionRow(icon: "pencil",
                      title: String(localized: "common.edit", defaultValue: "Edit"),
                      tint: palette.ink2) {
                onEdit()
                dismissIfNeeded()
            }

            Spacer(minLength: 0)
        }
        .padding(.top, 8)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(palette.paper)
        .confirmationDialog(
            String(localized: "tickleAction.snooze.title", defaultValue: "Snooze for how long?"),
            isPresented: $showingSnoozeOptions,
            titleVisibility: .visible
        ) {
            Button(String(localized: "tickleAction.snooze.oneDay", defaultValue: "1 day")) {
                onSnooze(1)
                dismissIfNeeded()
            }
            Button(String(localized: "tickleAction.snooze.threeDays", defaultValue: "3 days")) {
                onSnooze(3)
                dismissIfNeeded()
            }
            Button(String(localized: "tickleAction.snooze.oneWeek", defaultValue: "1 week")) {
                onSnooze(7)
                dismissIfNeeded()
            }
        }
    }

    private var header: some View {
        HStack(spacing: 12) {
            avatar
            VStack(alignment: .leading, spacing: 2) {
                Text(displayName)
                    .font(.system(size: 18, weight: .semibold))
                    .foregroundStyle(palette.ink)
                    .lineLimit(1)
                Text(reminder.frequency.localizedName)
                    .font(.system(size: 13))
                    .foregroundStyle(palette.ink2)
            }
            Spacer()
        }
        .padding(.horizontal, WarmSpacing.lg)
        .padding(.bottom, 14)
    }

    @ViewBuilder
    private var avatar: some View {
        if let contact = reminder.contact {
            ContactPhotoView(contact: contact, category: category, style: .list)
        } else {
            ZStack {
                Circle()
                    .fill(category.palette.accent)
                    .frame(width: 36, height: 36)
                if reminder.group != nil {
                    Text(reminder.group?.emoji ?? "👥").font(.system(size: 18))
                } else {
                    Image(systemName: "bell.fill")
                        .foregroundStyle(Color(red: 0.98, green: 0.96, blue: 0.89))
                }
            }
        }
    }

    private func actionRow(icon: String,
                           title: String,
                           tint: Color,
                           action: @escaping () -> Void) -> some View {
        Button(action: action) {
            HStack(spacing: 14) {
                Image(systemName: icon)
                    .font(.system(size: 16, weight: .semibold))
                    .foregroundStyle(tint)
                    .frame(width: 38, height: 38)
                    .background(tint.opacity(0.14))
                    .clipShape(Circle())
                Text(title)
                    .font(.system(size: 16, weight: .medium))
                    .foregroundStyle(palette.ink)
                Spacer()
            }
            .padding(.horizontal, WarmSpacing.lg)
            .padding(.vertical, 9)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }
}

File: android/app/src/main/java/com/xaymaca/sit/ui/tickle/TickleActionSheet.kt
package com.xaymaca.sit.ui.tickle

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Message
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Edit
import androidx.compose.material.icons.filled.Email
import androidx.compose.material.icons.filled.Phone
import androidx.compose.material.icons.filled.Snooze
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.xaymaca.sit.R
import com.xaymaca.sit.data.model.TickleFrequency
import com.xaymaca.sit.ui.shared.displayNameResId
import com.xaymaca.sit.ui.theme.WarmCategory
import com.xaymaca.sit.ui.theme.WarmSpacing
import com.xaymaca.sit.ui.theme.WarmTheme

/**
 * Connect-action sheet shown when a tickle row is tapped (TIC-36). Surfaces the
 * contact's available channels (Message / Call / Email) plus the secondary tickle
 * actions (Mark done / Snooze / Edit). Channel rows are hidden when the tickle has
 * no contact (group tickles) or the contact lacks that channel.
 *
 * Every action hands off to a system composer/dialer — no contact data leaves the
 * device. The hosting screen owns intent dispatch and dismissal.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun TickleActionSheet(
    target: TickleViewModel.TickleActionTarget,
    onCompose: () -> Unit,
    onCall: () -> Unit,
    onEmail: () -> Unit,
    onMarkDone: () -> Unit,
    onSnooze: () -> Unit,
    onEdit: () -> Unit,
    onDismiss: () -> Unit,
) {
    val palette = WarmTheme.palette()
    val category = WarmCategory.from(target.categoryId) ?: WarmCategory.Community

    ModalBottomSheet(onDismissRequest = onDismiss, containerColor = palette.paper) {
        Column(modifier = Modifier.fillMaxWidth()) {
            // Header: avatar + name + frequency
            Row(
                modifier = Modifier.fillMaxWidth().padding(horizontal = WarmSpacing.Lg, vertical = 4.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Box(
                    modifier = Modifier.size(40.dp).clip(CircleShape).background(category.palette.accent),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        text = target.initials,
                        style = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.SemiBold, color = Color(0xFFFAF4E2)),
                    )
                }
                Spacer(modifier = Modifier.width(12.dp))
                Column {
                    Text(
                        text = target.displayName,
                        style = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.SemiBold, color = palette.ink),
                        maxLines = 1,
                    )
                    Text(
                        text = stringResource(TickleFrequency.valueOf(target.reminder.frequency).displayNameResId),
                        style = TextStyle(fontSize = 13.sp, color = palette.ink2),
                    )
                }
            }

            Spacer(modifier = Modifier.height(8.dp))

            if (target.phones.isNotEmpty()) {
                ActionRow(Icons.AutoMirrored.Filled.Message, stringResource(R.string.tickle_action_compose), category.palette.accent, onCompose)
                ActionRow(Icons.Default.Phone, stringResource(R.string.warm_contact_call), category.palette.accent, onCall)
            }
            if (target.emails.isNotEmpty()) {
                ActionRow(Icons.Default.Email, stringResource(R.string.contact_detail_section_email), category.palette.accent, onEmail)
            }

            Box(modifier = Modifier.fillMaxWidth().padding(start = 64.dp).height(1.dp).background(palette.cardBorder))

            ActionRow(Icons.Default.CheckCircle, stringResource(R.string.tickle_row_action_done), palette.ink2, onMarkDone)
            ActionRow(Icons.Default.Snooze, stringResource(R.string.tickle_row_action_snooze), palette.ink2, onSnooze)
            ActionRow(Icons.Default.Edit, stringResource(R.string.common_edit), palette.ink2, onEdit)

            Spacer(modifier = Modifier.height(24.dp))
        }
    }
}

@Composable
private fun ActionRow(
    icon: ImageVector,
    label: String,
    tint: Color,
    onClick: () -> Unit,
) {
    val palette = WarmTheme.palette()
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(horizontal = WarmSpacing.Lg, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Box(
            modifier = Modifier.size(38.dp).clip(CircleShape).background(tint.copy(alpha = 0.14f)),
            contentAlignment = Alignment.Center,
        ) {
            Icon(icon, contentDescription = null, tint = tint, modifier = Modifier.size(18.dp))
        }
        Text(
            text = label,
            style = TextStyle(fontSize = 16.sp, fontWeight = FontWeight.Medium, color = palette.ink),
        )
    }
}
